improve error message for cost function

amanmdesai opened the pull request:
Hello @HDembinski

This PR is regarding https://github.com/scikit-hep/iminuit/issues/791

HDembinski commented:
Cool, thanks! I think the error message is spot on.

Would you be willing to write a test for this? Otherwise I would merge as is and write a test at some later point.

A test would go to `tests/test_cost.py` and the pytest facility to use is `with pytest.raises(ValueError): <create condition which triggers the exception>`. Please have a look at examples in the source code, just search for `pytest.raises`.

HDembinski commented:
Thinking more about it, there are a couple of things that need to be improved.

- I think we should check that the shapes are compatible, not only the length, since the binned cost functions support N-dimensional histograms.
- The check should be implemented in class `BinnedCostWithModel`, so that all derived classes benefit from it. A good place seems to be `BinnedCostWithModel._pred`.

amanmdesai commented:
Hi @HDembinski 

I have added a test for raisevalue error using pytest.raises. However this test fails since scipy is not installed. 
In some cases, the other tests in test_cost use the following statement:

```
@pytest.mark.skipif(not scipy_available, reason="scipy.stats is needed")
```

so shall I include the above statement to skip test if scipy is not available?
locally, the test seem to pass.

I have also implemented a check for shape compatibility. 

Please let me know if something needs to be changed.

Thanks,
Aman

HDembinski commented:
> so shall I include the above statement to skip test if scipy is not available?

Yes, please. This is happening because scipy is an optional dependency in some methods. Those methods work with and without scipy being installed, but execute different code in both cases. We want to test both code paths, that's why the tests are run with and without scipy. When you test a method for which scipy is required, you need to explicitly mark so that it is skipped in those test runs.

amanmdesai commented:
> > so shall I include the above statement to skip test if scipy is not available?
> 
> Yes, please. This is happening because scipy is an optional dependency in some methods. Those methods work with and without scipy being installed, but execute different code in both cases. We want to test both code paths, that's why the tests are run with and without scipy. When you test a method for which scipy is required, you need to explicitly mark so that it is skipped in those test runs.

I see. Thanks. I have added that statment.
I am not sure what to write in tests to check compatibility of shapes (which is probably the reason for decrease in coverage).

HDembinski commented on src/iminuit/cost.py:
```diff
@@ -1126,12 +1126,12 @@ class BinnedCostWithModel(BinnedCost):
 
-    __slots__ = "_xe_shape", "_model", "_model_xe"
+    __slots__ = "_xe_shape", "_model", "_model_xe", "_n"
 
     def __init__(self, n, xe, model, verbose):
         """For internal use."""
         self._model = model
-
+        self._n = n
```
It is not necessary to add `_n` to the slots. If you follow the inheritance chain upwards to `MaskedClost`, you can see that the data array `n` is stored in the slot `_data`. So you can query `self._data` instead of `self._n`

amanmdesai replied:
Thanks. Done.

HDembinski commented on src/iminuit/cost.py:
```diff
@@ -1142,6 +1142,8 @@ def __init__(self, n, xe, model, verbose):
             self._model_xe = np.row_stack(
                 [x.flatten() for x in np.meshgrid(*self.xe, indexing="ij")]
             )
+            if len(self._xe_shape) != len(self._model_xe):
+                raise ValueError("Variable shapes do not match")
```
I think this always matches by construction, no?

HDembinski commented on src/iminuit/cost.py:
```diff
@@ -1153,6 +1155,11 @@ def _pred(self, args: Sequence[float]) -> NDArray:
         # differences can come out negative due to round-off error in subtraction,
         # we set negative values to zero
         d[d < 0] = 0
+        if len(self._n) != len(d):
```
Here you can use `self._data`.

HDembinski commented on src/iminuit/cost.py:
```diff
@@ -1153,6 +1154,11 @@ def _pred(self, args: Sequence[float]) -> NDArray:
         # differences can come out negative due to round-off error in subtraction,
         # we set negative values to zero
         d[d < 0] = 0
+        if len(self._data) != len(d):
+            raise ValueError(
+                f"Expected model to return an array of size {len(self._data)},"
+                f"but it returns an array of size {len(d)}"
+            )
```
Good, but `d` and `self._data` could be multidimensional, so we want to check `self._data.shape == d.shape`.

HDembinski replied:
The error message needs to be adapted according, "array of size" -> "array with shape" etc

HDembinski commented on tests/test_cost.py:
```diff
@@ -1398,3 +1398,24 @@ def test_deprecated_Template_method():
         t._impl is cost.template_chi2_da
+
+
+@pytest.mark.skipif(not scipy_available, reason="scipy.stats is needed")
+def test_error_message_cost():
+    from iminuit import cost
+    import numpy as np
+    from scipy.stats import norm
```
Please also test the multi-dimensional case, where `edges` is a tuple of two arrays, with edges for the x and y-axis, and n is a 2d array.

amanmdesai commented:
Hi @HDembinski 

I have implemented all suggesstions with the exception of 2D test (I could not implement this test).

HDembinski commented:
> I have implemented all suggesstions with the exception of 2D test (I could not implement this test).

Ok, I implemented the 2D test and found some issue with our implementation. We should already test the length of the array returned by the model before it is further reshaped by `_pred`, otherwise there still will be confusing error messages.

I made a couple of unrelated changes, to make mypy happier and to increase coverage back to 100 %.

amanmdesai commented:
> > I have implemented all suggesstions with the exception of 2D test (I could not implement this test).
> 
> Ok, I implemented the 2D test and found some issue with our implementation. We should already test the length of the array returned by the model before it is further reshaped by `_pred`, otherwise there still will be confusing error messages.
> 
> I made a couple of unrelated changes, to make mypy happier and to increase coverage back to 100 %.

Thanks a lot @HDembinski for your help and guidance!

HDembinski commented:
Thanks @amanmdesai !
